File: recipes/vectors/verify.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# verify.rb — check an xenv decrypt implementation against the official
# conformance vectors in ./vectors.json, using NOTHING but this file and
# that JSON. No vault, no `xenv` binary, no network.
#
# This doubles as the smallest possible reference read-only loader: given
# (passphrase, salt, iter, envelope) it returns plaintext bytes. Port these
# ~20 lines of crypto to any language and run it against vectors.json to
# prove your loader is correct — that is the whole conformance contract.
#
#   ruby verify.rb            # → prints PASS/FAIL per vector, exits non-zero on any failure

require "openssl"
require "json"
require "base64"

VAULT_VERSION = "v3"

# (passphrase, salt-hex, iter) -> [enc_key, mac_key]
def derive_keys(pass, salt_hex, iter)
  salt = [salt_hex].pack("H*")
  out  = OpenSSL::PKCS5.pbkdf2_hmac(pass, salt, iter, 64, OpenSSL::Digest.new("SHA256"))
  [out[0, 32], out[32, 32]]
end

# raw envelope string -> plaintext bytes (raises on any tamper / bad key)
def decrypt(envelope, enc_key, mac_key)
  tag, ver, iv_hex, ct_hex, mac_hex, extra = envelope.strip.split(":")
  raise "envelope: wrong field count" if extra || mac_hex.nil?
  raise "envelope: unsupported #{tag}:#{ver}" unless tag == "xenv" && ver == VAULT_VERSION
  raise "envelope: wrong iv/mac length" unless iv_hex.length == 32 && mac_hex.length == 64
  raise "envelope: ct not block-aligned" if ct_hex.empty? || ct_hex.length % 32 != 0
  raise "envelope: non-hex" unless (iv_hex + ct_hex + mac_hex).match?(/\A[0-9a-f]+\z/)

  # MAC verify FIRST — encrypt-then-MAC, constant-time, scope binds version+iv+ct.
  scope    = "#{VAULT_VERSION}:#{iv_hex}:#{ct_hex}"
  expected = OpenSSL::HMAC.digest("SHA256", mac_key, scope)
  provided = [mac_hex].pack("H*")
  unless expected.bytesize == provided.bytesize &&
         OpenSSL.fixed_length_secure_compare(expected, provided)
    raise "MAC verification failed"
  end

  c = OpenSSL::Cipher.new("aes-256-cbc")
  c.decrypt
  c.key = enc_key
  c.iv  = [iv_hex].pack("H*")
  c.update([ct_hex].pack("H*")) + c.final
end

data = JSON.parse(File.read(File.join(__dir__, "vectors.json")))
pass = data.fetch("passphrase")
fail_count = 0

data.fetch("vectors").each do |v|
  enc, mac = derive_keys(pass, v.fetch("salt"), v.fetch("iter"))
  label = "#{v['name']} (#{v['expect']})"
  begin
    got = decrypt(v.fetch("envelope"), enc, mac)
    if v["expect"] == "ok"
      want = Base64.decode64(v.fetch("plaintext_b64"))
      if got == want
        puts "  ok    #{label}"
      else
        puts "  FAIL  #{label}: plaintext mismatch"
        fail_count += 1
      end
    else # expected a failure but decrypt succeeded
      puts "  FAIL  #{label}: expected #{v['expect']} but decrypt SUCCEEDED"
      fail_count += 1
    end
  rescue StandardError => e
    if v["expect"] == "ok"
      puts "  FAIL  #{label}: #{e.message}"
      fail_count += 1
    else
      puts "  ok    #{label}: rejected (#{e.message})"
    end
  end
end

puts(fail_count.zero? ? "\nALL VECTORS PASS" : "\n#{fail_count} FAILED")
exit(fail_count.zero? ? 0 : 1)

File: recipes/vectors/verify.js
#!/usr/bin/env node
// verify.js — the Node twin of verify.rb. Checks a decrypt implementation
// against ./vectors.json using nothing but this file + that JSON. Proves the
// vectors are a language-neutral conformance oracle: same JSON, any language.
//
//   node verify.js     # PASS/FAIL per vector; non-zero exit on any failure

const crypto = require("crypto");
const fs = require("fs");
const path = require("path");

function deriveKeys(pass, saltHex, iter) {
  const out = crypto.pbkdf2Sync(pass, Buffer.from(saltHex, "hex"), Number(iter), 64, "sha256");
  return { encKey: out.subarray(0, 32), macKey: out.subarray(32, 64) };
}

// Dual-read: v3 takes salt/iter from the caller; v4 is self-contained.
function decrypt(envelope, passphrase, v3Salt, v3Iter) {
  const parts = envelope.trim().split(":");
  if (parts[0] !== "xenv") throw new Error("envelope: not xenv");

  let saltHex, iter, ivHex, ctHex, macHex, scope, keys;
  if (parts[1] === "v3") {
    if (parts.length !== 5) throw new Error("envelope: wrong field count");
    [, , ivHex, ctHex, macHex] = parts;
    saltHex = v3Salt; iter = v3Iter;
    scope = `v3:${ivHex}:${ctHex}`;
  } else if (parts[1] === "v4") {
    if (parts.length !== 7) throw new Error("envelope: wrong field count");
    [, , saltHex, iter, ivHex, ctHex, macHex] = parts;
    if (!/^[0-9a-f]{32}$/.test(saltHex)) throw new Error("envelope: bad salt");
    // iter is attacker-controllable in v4 → bound it before PBKDF2 (DoS guard)
    if (!/^[0-9]+$/.test(iter) || Number(iter) < 1 || Number(iter) > 10000000) throw new Error("envelope: bad iter");
    scope = `v4:${saltHex}:${iter}:${ivHex}:${ctHex}`;
  } else if (parts[1] === "v5") {
    if (parts.length !== 8) throw new Error("envelope: wrong field count");
    let valueSalt;
    [, , saltHex, iter, valueSalt, ivHex, ctHex, macHex] = parts;
    if (!/^[0-9a-f]{32}$/.test(saltHex) || !/^[0-9a-f]{32}$/.test(valueSalt)) throw new Error("envelope: bad salt");
    if (!/^[0-9]+$/.test(iter) || Number(iter) < 1 || Number(iter) > 10000000) throw new Error("envelope: bad iter");
    // two-level KDF: PBKDF2 master over the shared salt, then HKDF per value.
    const master = crypto.pbkdf2Sync(passphrase, Buffer.from(saltHex, "hex"), Number(iter), 64, "sha256");
    const okm = Buffer.from(crypto.hkdfSync("sha256", master, Buffer.from(valueSalt, "hex"), Buffer.from("xenv:v5"), 64));
    keys = { encKey: okm.subarray(0, 32), macKey: okm.subarray(32, 64) };
    scope = `v5:${saltHex}:${iter}:${valueSalt}:${ivHex}:${ctHex}`;
  } else {
    throw new Error(`envelope: unsupported version ${parts[1]}`);
  }

  if (ivHex.length !== 32 || macHex.length !== 64) throw new Error("envelope: wrong iv/mac length");
  if (!ctHex || ctHex.length % 32 !== 0) throw new Error("envelope: ct not block-aligned");
  if (!/^[0-9a-f]+$/.test(saltHex + ivHex + ctHex + macHex)) throw new Error("envelope: non-hex");

  const { encKey, macKey } = keys || deriveKeys(passphrase, saltHex, iter);
  const expected = crypto.createHmac("sha256", macKey).update(scope, "ascii").digest();
  const provided = Buffer.from(macHex, "hex");
  if (expected.length !== provided.length || !crypto.timingSafeEqual(expected, provided)) {
    throw new Error("MAC verification failed");
  }
  const d = crypto.createDecipheriv("aes-256-cbc", encKey, Buffer.from(ivHex, "hex"));
  return Buffer.concat([d.update(Buffer.from(ctHex, "hex")), d.final()]);
}

const data = JSON.parse(fs.readFileSync(path.join(__dirname, "vectors.json"), "utf8"));
let fails = 0;

for (const v of data.vectors) {
  const label = `${v.name} (${v.wire} ${v.expect})`;
  try {
    const got = decrypt(v.envelope, data.passphrase, v.salt, v.iter);
    if (v.expect === "ok") {
      const want = Buffer.from(v.plaintext_b64, "base64");
      if (got.equals(want)) console.log(`  ok    ${label}`);
      else { console.log(`  FAIL  ${label}: plaintext mismatch`); fails++; }
    } else {
      console.log(`  FAIL  ${label}: expected ${v.expect} but decrypt SUCCEEDED`); fails++;
    }
  } catch (e) {
    if (v.expect === "ok") { console.log(`  FAIL  ${label}: ${e.message}`); fails++; }
    else console.log(`  ok    ${label}: rejected (${e.message})`);
  }
}

console.log(fails === 0 ? "\nALL VECTORS PASS" : `\n${fails} FAILED`);
process.exit(fails === 0 ? 0 : 1);
